### PR TITLE
DOCSP-40806 Indexes

### PR DESCRIPTION
**Pull request by jordan-smith721**
# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-java/blob/master/REVIEWING.md)

JIRA - https://jira.mongodb.org/browse/DOCSP-40806
Staging - https://preview-mongodbjordansmith721.gatsbyjs.io/cpp-driver/DOCSP-40806-indexes/work-with-indexes/

## Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Are all the links working?
- [x] Are the [facets and meta keywords](https://wiki.corp.mongodb.com/display/DE/Docs+Taxonomy) accurate?

### CODE REVIEW

File: source/work-with-indexes.txt [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+Operational Considerations
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To improve query performance, build indexes on fields that appear often in
+your application's queries and operations that return sorted results. Each
+index that you add consumes disk space and memory when active, so we recommend
+that you track index memory and disk usage for capacity planning. In addition,
+when a write operation updates an indexed field, MongoDB updates the related
+index.
```
Comment (mongoKart) [inline source/work-with-indexes.txt:41]:
```suggestion
when a write operation updates an indexed field, MongoDB also updates any related
indexes.
```

File: source/work-with-indexes.txt [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+MongoDB supports several different index types to support querying your data.
+The following sections describe the most common index types and provide sample
+code for creating each index type.
+
+Single-Field Indexes
+~~~~~~~~~~~~~~~~~~~~
+
+:manual:`Single-field indexes </core/index-single/>` are indexes with a
+reference to a single field within a collection's
+documents. They improve single field query and sort performance, and support
```
Comment (mongoKart) [inline source/work-with-indexes.txt:73]:
```suggestion
documents. They improve single-field query and sort performance, and support
```

File: source/work-with-indexes.txt [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+:manual:`TTL Indexes </core/index-ttl>` that
```
Comment (mongoKart) [inline source/work-with-indexes.txt:74]: Q: Is this describing just one kind of TTL index, or is it a definition? If the latter, I'd do this:

```suggestion
:manual:`TTL Indexes </core/index-ttl>`, which
```

Comment (mongoKart) [inline source/work-with-indexes.txt:74]:
```suggestion
:manual:`TTL (time to live) indexes </core/index-ttl>` that
```

File: source/work-with-indexes.txt [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+.. _cpp-indexes-remove:
+
+Remove an Index
+---------------
+
+You can remove any unused index except the default unique index on the
+``_id`` field.
+
+The following sections show how to remove a single index or to remove all
+indexes in a collection.
```
Comment (mongoKart) [inline source/work-with-indexes.txt:104]:
```suggestion
The following sections show how to remove a single index and how to remove all
indexes in a collection.
```

File: source/work-with-indexes.txt [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+For more information about designing your data model and choosing indexes appropriate for your application, see the
+:manual:`Data Modeling and Indexes </core/data-model-operations/#indexes>` guide
+in the {+mdb-server+} manual.
+
+Sample Data
+~~~~~~~~~~~
+
+The examples in this guide use the ``sample_mflix.movies`` collection
+from the :atlas:`Atlas sample datasets </sample-data>`. To learn how to create a
+free MongoDB Atlas cluster and load the sample datasets, see the
```
Comment (mongoKart) [inline source/work-with-indexes.txt:58]:
```suggestion
free MongoDB Atlas cluster and load the sample datasets, see
```

**Review by mongoKart — APPROVED** [review]
some small changes and suggestions, but lgtm otherwise

File: source/work-with-indexes.txt (hunk 3 above)
Comment (jordan-smith721) [reply source/work-with-indexes.txt]: I think it's the latter

File: source/work-with-indexes.txt [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+   You cannot remove a single field from a compound text index. You must
+   drop the entire index and create a new one to update the indexed
+   fields.
+
+Remove All Indexes
+~~~~~~~~~~~~~~~~~~
+
+Starting with MongoDB 4.2, you can drop all indexes by calling the
+``drop_indexes()`` method on your collection:
```
Comment (Julia-Garland) [inline source/work-with-indexes.txt:130]: The C++ driver doesn't have a drop_indexes() method that can be called on a collection.

Use the drop_all() method instead, which like drop_one() is called on an index_view to the collection, ex. `coll.indexes().drop_all()`

File: source/includes/indexes/indexes.cpp [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+auto index_specification = make_document(kvp("title", 1));
+collection.create_index(std::move(index_specification));
+// end-index-single
+
+// start-remove-index
+collection.drop_index("title_1");
+// end-remove-index
+
+// start-remove-all-indexes
+collection.drop_indexes();
```
Comment (Julia-Garland) [inline source/includes/indexes/indexes.cpp:11]: `collection.indexes().drop_all();`

File: source/includes/indexes/indexes.cpp [hunk 8]
```diff
@@ -0,0 +1,16 @@
+// start-index-single
+auto index_specification = make_document(kvp("title", 1));
+collection.create_index(std::move(index_specification));
+// end-index-single
+
+// start-remove-index
+collection.drop_index("title_1");
```
Comment (Julia-Garland) [inline source/includes/indexes/indexes.cpp:7]: `collection.indexes().drop_one("title_1");`

File: source/work-with-indexes.txt [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+``_id`` field.
+
+The following sections show how to remove a single index and how to remove all
+indexes in a collection.
+
+Remove a Single Index
+~~~~~~~~~~~~~~~~~~~~~
+
+Pass an instance of an index or the index name to the ``drop_index()`` method to
+remove an index from a collection.
```
Comment (Julia-Garland) [inline source/work-with-indexes.txt:110]: The C++ driver doesn't have a drop_index() method that can be called on a collection. 

Instead, you can obtain an index_view to the collection and call drop_one(), ex. `coll.indexes().drop_one(name)`

File: source/work-with-indexes.txt [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+To learn more about indexes in MongoDB, see the :manual:`Indexes </indexes>`
+guide in the {+mdb-server+} manual.
+
+API Documentation
+~~~~~~~~~~~~~~~~~
+
+To learn more about any of the methods or types discussed in this
+guide, see the following API documentation:
+
+- `create_index() <{+api+}/classmongocxx_1_1v__noabi_1_1collection.html#a39cf05fd8da3a7993929c8bfd3de9b46>`__
```
Comment (Julia-Garland) [inline source/work-with-indexes.txt:157]: Additional relevant links:
- [drop_all()](https://mongocxx.org/api/current/classmongocxx_1_1v__noabi_1_1index__view.html#a2fc4f2778ce800076368f026fd2649d8)
- [drop_one()](https://mongocxx.org/api/current/classmongocxx_1_1v__noabi_1_1index__view.html#a1779a23bd9565cf295cc2479b6e5981a)
- [indexes()](https://mongocxx.org/api/current/classmongocxx_1_1v__noabi_1_1collection.html#aac9843f8a560d39b85ef24a651d66e3b)

File: source/work-with-indexes.txt [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+Starting with MongoDB 4.2, you can drop all indexes by calling the
+``drop_indexes()`` method on your collection:
+
+.. literalinclude:: /includes/indexes/indexes.cpp
+   :language: cpp
+   :start-after: start-remove-all-indexes
+   :end-before: end-remove-all-indexes
+
+For earlier versions of MongoDB, pass ``"*"`` as a parameter to your call to
+``drop_index()`` on your collection:
```
Comment (Julia-Garland) [inline source/work-with-indexes.txt:138]: Replace drop_index() with drop_one(). Replace collection with "collection's index view"

File: source/includes/indexes/indexes.cpp [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+// end-remove-index
+
+// start-remove-all-indexes
+collection.drop_indexes();
+// end-remove-all-indexes
+
+// start-remove-all-wildcard
+collection.drop_index("*");
```
Comment (Julia-Garland) [inline source/includes/indexes/indexes.cpp:15]: `collection.indexes().drop_one("*");`

**Review by Julia-Garland — CHANGES_REQUESTED** [review]
I requested some changes so the page matches the C++ driver API. LGTM otherwise!

File: source/work-with-indexes.txt (hunk 6 above)
Comment (jordan-smith721) [reply source/work-with-indexes.txt]: Ah, thanks for pointing this out! Updated all samples

File: source/work-with-indexes.txt (hunk 10 above)
Comment (jordan-smith721) [reply source/work-with-indexes.txt:157]: Thanks for finding these! I was struggling (probably because I was looking up the wrong method name)

**Review by Julia-Garland — APPROVED** [review]
LGTM!

**Pull request merged**